fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -179,69 +179,126 @@ private IndexingByIndex getIndexerByIndex() {
+                        .addKeysAndValues(common.indexLogMessageKeyValues())
+                        .toString());
+            }
+            throw FDBExceptions.wrapException(ex);
+        }
+
+        indexer = null; // clear cache - avoid reuse of the same indexer
+        final IndexStatePrecondition indexStatePrecondition = common.getIndexStatePrecondition();
+        final IndexingBase.PartlyBuiltException partlyBuiltException = IndexingBase.getAPartlyBuildExceptionIfApplicable(ex);
         if (partlyBuiltException != null) {
+            // An ongoing indexing process with a different method type was found. Some precondition cases should be handled.
             IndexBuildProto.IndexBuildIndexingStamp conflictingIndexingTypeStamp = partlyBuiltException.savedStamp;
-            // Here: an indexing type stamp already exists, with a different type. Few precondition cases are handled
             if (LOGGER.isInfoEnabled()) {
                 LOGGER.info(KeyValueLogMessage.build("conflicting indexing type stamp",
-                        LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
-                        LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
-                        LogMessageKeys.INDEXER_ID, common.getUuid(),
-                        LogMessageKeys.INDEX_STATE_PRECONDITION, common.getIndexStatePrecondition(),
-                        LogMessageKeys.ACTUAL_TYPE, conflictingIndexingTypeStamp
-                ).toString());
+                        LogMessageKeys.CURR_ATTEMPT, attemptCount,
+                        LogMessageKeys.INDEX_STATE_PRECONDITION, indexStatePrecondition,
+                        LogMessageKeys.ACTUAL_TYPE, conflictingIndexingTypeStamp)
+                        .addKeysAndValues(common.indexLogMessageKeyValues())
+                        .toString());
             }
-            if (common.getIndexStatePrecondition() == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY) {
+
+            if (indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY) {
+                // Make best effort to finish indexing. Attempt continuation of the previous method
                 indexer = null;
-                // now, match the policy to the previous run
+                // Now, match the policy to the previous run
                 IndexBuildProto.IndexBuildIndexingStamp.Method method = conflictingIndexingTypeStamp.getMethod();
-                if ( method == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+                if (method == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+                    // Partly built by records. The fallback indicator should handle the policy
                     fallbackToRecordsScan = true;
-                    return fallbackFunction.get();
+                    return indexingLauncher(indexingFunc, attemptCount);
```

Comment:
I see you're preserving the `indexingFunc` here and in every method call. If I understand the approach, the idea is that the `indexingFunc` method implementations look at state in the `OnlineIndexer` and react to that. So, for example, if the initial index was by-index, but then it had to fallback to a by-records scan, now calling the `indexerFunc` again which reads that the appropriate boolean has been flipped and that this uses the other indexer. Is that the idea?

Not sure if that should be documented somehow, or the right way to document that.

Reply:
Yes, you are right (of course:). 
At the beginning of the recovery code there is the line: 
```  indexer = null; // clear cache - avoid reuse of the same indexer ```
Which enforces the next function call building a new indexer according to the modified parameters. I'll add this to the comment to clarify.